### DISCUSSION TITLE
Why is the right-click empty ?

### DISCUSSION THREAD

**Discussion by Yooops** [tannerhelland/PhotoDemon#459] [Q&A] [answered]
Hello PhotoDemon Community !

'New user' here (and not a native-english speaker as you'll see :) )
Well, I've just tried PhotoDemon when i see this new 9.0 version. What a change since the last time i gave it a try (the last one was... 6.2) !
I'm always looking for a simple, fast and useful software to easily modify some photos (crop, make a symetry, delete/modify a part of the picture, ...). I'm not a huge fan of the big ones as Paint.NET or GIMP that i find too hard to handle for the newcomer.

My first question (for now) is in the subjet of this post : why is the context-menu empty ?

In my 'everyday' photo-software (i find it extremely useful, but the project seem unfortunately discontinued, and the UI is so WinXP-like that i'd like to find a new one), the right-click give me some choices as 'crop to selection', 'layer via copy', 'paste to new image', ...
These fast choices are time-saving and very user-friendly.

Is this just a dev choice ? or is it on the TODO list ?

Thx a lot for all your work on this beautiful project.

**Comment by tannerhelland** [comment, marked as answer]
Hello @Yooops !  I apologize for the delay in replying.  Last week was a holiday week here and I unfortunately did not have much time for PhotoDemon.

Here's the short answer to your question: I would love PhotoDemon to have better right-click context menu behavior!  I completely agree that its current lack of context menus is unacceptable.  Can you tell me the name of (or link me to) the 'everyday' photo software you like?  I am always curious how other developers handle things like context menus and I learn a lot from studying other implementations.

Here is a longer answer to your question.  PhotoDemon is written in an ancient programming language (VB6, 25 years old next year!).  This "feature" is what allows PhotoDemon to run on a USB drive on any version of Windows - because it doesn't need access to any low-level Windows features newer than 1998.

This approach has obvious trade-offs.  One of my biggest problems is that the default UI of programs written in VB6 looks like they have time-traveled from 1998.  They also do not (easily) support languages other than the default language of the developer (English, in my case).

To overcome these problems, I wrote a custom UI toolkit specifically for PhotoDemon.  This took several years, but it let me add lacking features like Unicode support, dark visual themes, modern controls like sliders and much more.  Of course, this also meant I had to create every type of UI element required by a modern program - buttons, lists, sliders, etc.

The final UI element I need to custom-build is menus.  You can see this in PhotoDemon's main menu which just uses the default theme of whatever Windows OS it runs on.  The work required to implement menus is very large, larger than other UI elements, so I have put it off until the very end.  (This includes context menus, which also use normal Windows menus at present.)

So I have avoided implementing context menus yet because they will all need to be rewritten using the new custom menu system... whenever I am able to tackle that.  Once I have my own menu engine available, I expect to add an "explosion" of context menus throughout the program!

That is the long answer on why context menus are currently lacking in PhotoDemon.  I hope to tackle a new menu system sometime in the next year.  (It will require completely gutting the current menu system, so it is not a project I take lightly - the whole project depends on it!)

Thank you for the kind words and feedback, and I hope I can improve PhotoDemon's menus for you soon!

**Reply by Yooops** [reply]
Hello @tannerhelland,

Thank you for the long answer :) . I see now that it's a (big) technical difficulty and not a definitive point of view.

About what i called 'my-everyday-soft', i think i haven't used the good word : i spoke about 'project', but it was in the sense of 'life software cycle' and not github- or sourceforge-project. It was a freeware, that is becomed a shareware (old versions stay free but are not open-source, unfortunatly). And i find that the shareware version is now too complex (a great work, but that's not what i'm looking for). Just in case (but i don't know to what it could lead), it's called Photofiltre.

Thank you for what you've accomplished.
Even if there are some 'missing' tips or shortcuts, i have to say that PhotoDemon9 is one of the best project i've seen so far (really!). As i told above, i'm looking for a software with 'simple' UI and tools ('simple' = efficient, fast, useful, 'straight-to-the-job'), and for now, be sure PhotoDemon is certainly one of them !

**Comment by Mnewebmaster** [comment]
The program is superb. I hope right click functionality will arrive soon. Thank you
